budget/app.py

```diff
@@ -1,14 +1,176 @@
 import json
 import logging
 
+import boto3
+from botocore.exceptions import ClientError
+import requests
+
 log = logging.getLogger(__name__)
 log.setLevel(logging.DEBUG)
 
+# TODO configure this URL?
+SYNAPSE_TEAM_MEMBER_URL = 'https://repo-prod.prod.sagebase.org/repo/v1/teamMembers'
+
+# TODO pull this config from S3?
+USER_BUDGET_BY_TEAM = {
+  '12345': {
+    'amount': 2000,
+    'period': 'annual'
+  },
+  '67890': {
+    'amount': 100,
+    'period': 'annual'
+  }
+}
+
+
+def get_client(service):
+  return boto3.client(service)
+
+
+def get_users_by_team(teams):
+  '''Get users from synapse teams
+
+  Returns a dictionary of users with a list of their team memberships
+  '''
+  teams_by_user_id = {}
+  for team_id in teams:
+    synapse_url = f'{SYNAPSE_TEAM_MEMBER_URL}/{team_id}'
+    response = requests.get(synapse_url)
+    if response.ok:
+      results = json.loads(response.text)['results']
+      user_ids = [result['member']['ownerId'] for result in results if not result['isAdmin']]
+    else:
+      response.raise_for_status()
+    for user_id in user_ids:
+      if user_id in teams_by_user_id:
+        teams_by_user_id[user_id].append(team_id)
+      else:
+        teams_by_user_id[user_id] = [team_id]
+  return teams_by_user_id
+
+
+def check_user_duplicates(teams_by_user_id):
+  '''Verify that no users occur in multiple teams'''
+
+  duplicates = {user_id:team_memberships
+    for (user_id,team_memberships) in teams_by_user_id.items()
+    if len(team_memberships) > 1
+  }
+
+  if not duplicates:
+    return ''  # return an empty message if there are no duplicates
+  else:
+    return '\n'.join([
+      f'Synapse user id {user_id} occurs in teams {", ".join(team_memberships)}'
+      for (user_id, team_memberships) in duplicates.items()
+    ])
+
+
+def compare_budgets_and_users(account_id, users):
+  '''Finds users who lack a budget
+
+  This checks budget names, which will use the format
+  f'service-catalog_{synapse_id}', against the user list,
+  returning a list of user ids that need a budget to be made
+  '''
+  budgets_client = get_client('budgets')
+
+  # get budgets
+  budgets = budgets_client.describe_budgets(
+    AccountId=account_id
+    ).get('Budgets')
+
+  # get just the budget names
+  service_catalog_budget_names = [
+    budget['BudgetName'] for budget in budgets
+    if budget['BudgetName'].startswith('service-catalog_')
```

Comment:
define a constant for 'service-catalog_', since you will use it in multiple places.

Reply:
👍